libs/langchain-google-common/src/utils/media_core.ts

```diff
@@ -0,0 +1,465 @@
+  }
+
+  mdelete(keys: string[]): Promise<void> {
+    return this.backingStore.mdelete(keys);
+  }
+
+  mget(keys: string[]): Promise<(MediaBlob | undefined)[]> {
+    return this.backingStore.mget(keys);
+  }
+
+  mset(keyValuePairs: [string, MediaBlob][]): Promise<void> {
+    return this.backingStore.mset(keyValuePairs);
+  }
+
+  yieldKeys(prefix: string | undefined): AsyncGenerator<string> {
+    return this.backingStore.yieldKeys(prefix);
+  }
+
+}
+
+export class SimpleWebBlobStore extends BlobStore {
+  _notImplementedException() {
+    throw new Error("Not implemented for SimpleWebBlobStore");
+  }
+
+  _hasValidPath(blob: MediaBlob, _opts?: BlobStoreStoreOptions): Promise<boolean> {
+    const {path} = blob;
+    if (path) {
+      const ret = path?.startsWith("http://") || path?.startsWith("https://");
+      return Promise.resolve(ret);
+    } else {
+      return Promise.resolve(false);
+    }
+  }
+
+  async _fetch(url: string): Promise<MediaBlob | undefined> {
+    const ret = new MediaBlob({
+      path: url,
+    });
+    const metadata: Record<string, unknown> = {};
+    const fetchOptions = {
+      method: "GET",
+    };
+    const res = await fetch(url, fetchOptions);
+    metadata.status = res.status;
+
+    const headers: Record<string, string> = {};
+    for (const [key, value] of res.headers.entries()) {
+      headers[key] = value;
+    }
+    metadata.headers = headers;
+
+    metadata.ok = res.ok;
+    if (res.ok) {
+      ret.data = await res.blob();
+    }
+
+    ret.metadata = metadata;
+    return ret;
+  }
+
+  async mget(keys: string[]): Promise<(MediaBlob | undefined)[]> {
+    const blobMap = keys.map(this._fetch);
+    return await Promise.all(blobMap);
+  }
+
+  async mdelete(_keys: string[]): Promise<void> {
+    this._notImplementedException();
+  }
+
+  async mset(_keyValuePairs: [string, MediaBlob][]): Promise<void> {
+    this._notImplementedException();
+  }
+
+  async *yieldKeys(_prefix: string | undefined): AsyncGenerator<string> {
+    this._notImplementedException();
+    yield "";
+  }
+}
+
+export interface MediaManagerConfiguration {
+  /**
+   * A map from the alias name to the canonical MediaBlob for that name
+   */
+  aliasStore: BlobStore;
+
+  /**
+   * The definitive store for the MediaBlob
+   */
+  canonicalStore: BlobStore;
+
+  /**
+   * BlobStore that can resolve a URL into the MediaBlob to save
+   * in the canonical store. A SimpleWebBlobStore is used if not provided.
+   */
+  resolver?: BlobStore;
+}
+
+/**
+ * Responsible for converting a URI (typically a web URL) into a MediaBlob.
+ * Allows for aliasing / caching of the requested URI and what it resolves to.
+ * This MediaBlob is expected to be usable to provide to an LLM, either
+ * through the Base64 of the media or through a canonical URI that the LLM
+ * supports.
+ */
+export class MediaManager {
+
+  aliasStore: BlobStore;
+
+  canonicalStore: BlobStore;
+
+  resolver: BlobStore;
```

Comment:
A bit more docs on why three stores are necessary would be good

Reply:
Each one is there to address different use cases of the `MediaManager`.
* The `resolver` is so a developer can include a generic URI that doesn't depend on the platform this runs on and get the data and metadata for this.
  * The most common case is providing an HTTP or HTTPS URL.
  * Having a resolver that deals with data: URIs also makes sense.
 * The `canonicalStore` is the store that takes the data (and possibly metadata) and saves it in a platform-dependent way.
  * So folks using AI Studio API would store it in the AI Studio File API. Those who are using Vertex AI would store it in Google Cloud Storage. There is a different BlobStore for each of these, and this is what would be set here.
  * The `MediaBlob` that this returns is the URI that should be sent to the LLM.
* The `aliasStore` should provide a long-term mapping between the two. 
  * So given a generic URI (ie - the HTTPS URL) for an image, get the `MediaBlob` that was saved for this URL already
  * This is important if you're using the URL for a video - you don't want to re-download and then re-upload the entire video every time.

The main use cases:
* The first time through in most interactions, you'll provide the public URL for the image: https://example.com/butterfly.jpg
  * The `resolver` is used to get the image itself
  * The image is handed to the `canonicalStore` that stores it and returns `MediaBlob` that has the path for how to reference it in the platform-specific store
  * This `MediaBlob` is stored in the `aliasStore` using the key of the public URL.
  * The LLM API uses the path from the `MediaBlob` that came from the `canonicalStore`
* Future times through
  * The `aliasStore` is used to get the `MediaBlob` for the public URL
  * If this isn't valid for some reason, it treats it like the first time through
  * If this is valid, the LLM API uses the path from the `MediaBlob` that came from the `aliasStore`

The advantage is that for both cases, all the developer has to think about is what the appropriate store classes should be for their project and then sending a public URL. The `MediaManager` class "does the right thing". 
* For images, the Python class always downloads the contents at the public URL to include it inline.
* For media, the current implementation (in both Python and JS) requires the developer to make sure the media has been uploaded correctly already and to provide the equivalent of the URL from the canonical store, rather than the public URL.

We can also evaluate some less common use cases (where the URL format from the canonical store is used, for example) to see how these might be supported, but currently those are handled using the legacy support that the LLM already supports.

Going forward, it might make sense to think about how to take a `MediaBlob` directly (for locally generated images, for example, or images downloaded from an image generator), but I'm not sure we have a good way to do that in general now.

Hopefully that is clearer, but I'm happy to discuss further to make sure we've identified all the critical cases.